src: fix generation of path objects in Windows

yamachu opened the pull request:
<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

fix: https://github.com/nodejs/node/issues/56650
ref: https://github.com/nodejs/node/pull/56657

This PR fixes a problem that caused a segmentation fault in module resolution when creating a require object with multibyte characters in a non-English environment.

Since the issue occurred when generating a std::filesystem::path object, some patches were applied to the changed areas in the following PR.

https://github.com/nodejs/node/commit/a7dad43d159df199e17bcc0c64e6f02a07fb7d9e

Since the issue was reproduced only in different locales, such as ja-JP on Windows, I rewrote the locale in the CI of coverage-windows to run the test.
https://github.com/yamachu/node/pull/2/files#diff-29094741d50149aa772b3e577ad509116bad722ad2de85689b6cb2c01e806a46

.github/workflows/coverage-windows.yml
```diff
+      - name: Change locale ja-JP for testing on SJIS environment
+        run: Set-WinSystemLocale -SystemLocale "ja-JP"
+      # to avoid configure, nobuild and noprojgen is needed
+      - name: Test on SJIS environment
+        run: ./vcbuild.bat nobuild noprojgen test-ci-js; node -e 'process.exit(0)'
+        env:
+          NODE_V8_COVERAGE: ./coverage/tmp
```

[The previous PR](https://github.com/nodejs/node/pull/56657) did not solve the problem when [using unicode](https://github.com/nodejs/node/blob/main/test/parallel/test-require-unicode.js), so the process was separated for windows and the path object was generated.

The part where std::filesystem::path is used can be fixed by changing this PR.

The results of the test conducted in a Japanese environment can be seen below.
https://github.com/yamachu/node/actions/runs/12903928231/job/35980111577#step:12:16844

anonrig reviewed (changes requested):
A similar already landed last August: https://github.com/nodejs/node/pull/54653

We should use `ToU8StringView()` rather than introducing a new API.

yamachu commented:
@anonrig 

Thanks for the review!
Does this mean in short that I should use std::u8string?

However, there is a problem with the u8string approach.
As I show in the description, the test-require-unicode.js test fails when I run code using u8string in an SJIS environment.

Therefore, I provided and used an API to handle wstrings without using ToU8StringView.

anonrig commented:
I recommend improving existing solution to fit both cases rather than implementing a new solution.

yamachu commented:
What are "both cases" presented here?

I am sure that ToU8StringView is not used in my current codebase.
Since it is not being used, it does not currently solve the problem.
It is not a solution because I know that using it again will cause problems.

I don't think that trying hard to use u8string in path is a better way to go....

yamachu commented:
The base branch is still bf59539b980cbada964bf4c0991afe55668526e8 , so the associated test seems to be fail. (experimental flag)
Should I rebase the branch and force push?

lpinca commented:
> Should I rebase the branch and force push?

Yes, thank you.

yamachu commented:
I had to rush to write my comments before work.
I apologize if any of my comments offended you.

I have rebased and force pushed the file, so please check with CI.

I would have liked to have been able to do this beforehand, but I looked over the entire code.
There I found the following helper functions.

https://github.com/nodejs/node/blob/d9786109b2a0982677135f0c146f6b591a0e4961/src/node_file.cc#L3149-L3160

This was exactly what I was looking for.
I replaced what I wrote in this PR with this and the test passed.

yamachu commented on src/node_modules.cc:
```diff
@@ -326,6 +330,22 @@ const BindingData::PackageConfig* BindingData::TraverseParent(
 
+#ifdef _WIN32
+std::wstring ConvertToWideString(const std::string& str) {
+  auto cp = GetACP();
```
This section was copied from node_file.cc, but this `GetACP` is particularly important.
As shown in the description, when the Windows locale is changed and executed, if the original UTF-8 code is left here, the strings cannot be handled properly.

lpinca replied:
I wonder if this utility function can be moved to `util-inl.h`? so that the same implementation that takes into account `GetACP` is used both here and in `node_file.cc` without duplication. I think this is also what @anonrig is requesting.

yamachu replied:
If we do the same for node_fs, test-fs-cp.mjs will falil...

It is very strange, but I have found that if I run the failed path(utf/~~~) in addition to the test-require-~ , it can be handled correctly.
It may be that the pre or post-process (if there is one) regarding path generation in node_fs is having an effect.

Therefore, this time I confined the implementation within node_modules instead of applying it to the whole system.

The following is the error log when applied to the entire system.
https://github.com/yamachu/node/actions/runs/12941644792/job/36098035649?pr=3#step:7:5033

code diff
https://github.com/yamachu/node/pull/3/commits/0041793aa35a6cf0c28d2334600f7e625439d77a

yamachu replied:
Ah, I see.
Do you mean to make these changes?

util-inl.h
```cpp
#ifdef _WIN32
inline std::wstring ConvertToWideString(const std::string& str, UINT cp) {
  int size_needed = MultiByteToWideChar(
      cp, 0, &str[0], static_cast<int>(str.size()), nullptr, 0);
  std::wstring wstrTo(size_needed, 0);
  MultiByteToWideChar(
      cp, 0, &str[0], static_cast<int>(str.size()), &wstrTo[0], size_needed);
  return wstrTo;
}
#endif  // _WIN32
```

node_file.cc
```cpp
#define BufferValueToPath(str)                                                 \
  std::filesystem::path(ConvertToWideString(str.ToString(), CP_UTF8))
```

I thought that this implementation would certainly have less impact and could be implemented with less code duplication.

lpinca replied:
Yes, exactly.

yamachu replied:
apply 👍  : https://github.com/nodejs/node/pull/56696/commits/4ddd7aeb532653367721f779e85be03746b955f7

yamachu commented:
https://github.com/yamachu/node/actions/runs/12933259601/job/36071441238#step:12:16876

Test has passed on SJIS Windows

lpinca reviewed (approved):
RSLGTM

lpinca commented:
Can you please squash the last 3 commits into the second or everything into one?

yamachu commented:
I squashed and added a test case for the newly found problem.

lpinca commented:
Can you please squash (last commit) again?

yamachu commented:
Squashed and force-pushed 👍

lpinca commented:
Landed in 97a3a8204c7c...176002400886.
